Fixing sql_connection.py module documentation

josenavas opened the pull request:
It also enables the doctests in this module.

Fixes #1356 
Fixes #982

antgonza commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,101 @@
+
+>>> from qiita_db.sql_connection import TRN
+
+Transaction blocks are used through the variable TRN available in this module.
```
Can you rephrase this as is not clear?

josenavas replied:
Any suggestion?

antgonza commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,101 @@
+>>> res
+[[[42]], [[43]], [[44]]]
+
+You can also use results from a previous command in the transaction in a later
```
Could you add a more complex example where we can se why we need the 3 ints?

antgonza commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,101 @@
+>>> res
+[[[42]], [[84]]]
+
+Multiple auxiliary functions exist for easy SQL result retrieval:
```
Is there a way to add headers? Cause I can see having separate sections:
- Transactions
- Querying
- Retrival
- Etc

josenavas replied:
I don't think we can add headers per se in here, as the headers are a controlled set in the documentation (e.g. here the header is "Examples"). However I can try to add a unordered list...

antgonza commented:
A few comments.

josenavas commented:
Comments addressed, thanks @antgonza

antgonza commented:
Any idea why the test failure?

josenavas commented:
The error test is in the doctests. I've activated them for this file, but I've been having problems running them locally, I will take a look and see if I can run them locally and fix them. The turn around time from travis is quite big...

josenavas commented:
I just fixed the doctest. I'm worried about one thing here. The tests are passing, however, these tests are not safe as the unit tests are, in the sense that they don't fail if the database is not a test database.

I have ideas on how to fix it, however, this is not possible until our code supports python 3.4. I need the [FAIL_FAST](https://docs.python.org/3/library/doctest.html#doctest.FAIL_FAST) flag for doctest, so after the first failure no other doctest is executed.

The first call in the doctests here destroys the database, as happens in the unittest. I think I should skip the doctests for now (by adding SKIP), but I would like to hear from other before doing it, as it is time consuming.

antgonza commented:
:+1: could you add an issue about the tests?

josenavas commented:
Sure, added here: https://github.com/biocore/qiita/issues/1415

@squirrelo @wasade @ElDeveloper anyone available for review?

wasade commented:
going through

wasade commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,139 @@
+
+>>> from qiita_db.sql_connection import TRN
+
+The `TRN` variable is an instance of the Transaction object. All queries should
```
Transaction -> `Transaction`

josenavas replied:
Done

wasade commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,139 @@
+>>> from qiita_db.sql_connection import TRN
+
+The `TRN` variable is an instance of the Transaction object. All queries should
+be executed through this transaction object to avoid committing anything to the
```
~~transaction~~

josenavas replied:
Done

wasade commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,139 @@
+
+The `TRN` variable is an instance of the Transaction object. All queries should
+be executed through this transaction object to avoid committing anything to the
+database that is partially executed. To perform any query you first need to
```
What about: "...to ensure atomicity." instead of "...to avoid committing ..."

josenavas replied:
Really good suggestion!

wasade commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,139 @@
+...     TRN.add("SELECT 42")
+...     res = TRN.execute()
+>>> res
+[[[42]]]
```
A curious developer may wonder why 3 layers of nesting...

josenavas replied:
Added a comment explaining the 3 levels of nesting.

wasade commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,139 @@
+>>> with TRN:
+...     TRN.add("SELECT 42")
+...     TRN.add("SELECT 43")
+...     res = TRN.execute_fetchindex(0)
```
It would be nice to explicitly note that the value `0` here corresponds to the first query in the transaction

josenavas replied:
Done

wasade commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,139 @@
+...     sql = ("SELECT * FROM qiita.preprocessed_data "
+...            "WHERE preprocessed_params_table = %s "
+...            "AND preprocessed_params_id = %s")
+...     TRN.add(sql, ["preprocessed_sequence_454_params", "{0:0:1}"])
```
The indexing here is not intuitive since the a `SELECT *` is being used. Possible to change the `SELECT *` to project only a specific column or columns?

josenavas replied:
@wasade the indexing is used to index the previous query (hence the 0 index) so it's actually selecting the `preprocessed_params_id` from the first row in the previous query

wasade replied:
Okay... so this is extremely confusing and not something that, in my opinion, is intuitive. I suggest adding a few more examples of the placeholders with detailed documentation working through the examples, and/or update the placeholders to be expressive.

wasade commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,139 @@
+You can also use results from a previous command in the transaction in a later
+command without partially executing the transaction by using placeholders. The
+placeholders are of the form {int:int:int}, where the first integer is the
+index of the query in the transaction, the second integer is the row in the
```
The row is difficult to understand since the cursor is not assured (AFAIK) to return rows in any particular order. In other words, the database is unordered unless a requested order is specified. I can see how this works when a single row is returned, but I can't help but thing anything besides `0` is meaningful unless the query includes an explicit ordering AND the developer knows in advance what the returned row of interest will be

josenavas replied:
You're correct, the order in the results is not assured unless a requested order is specified.

wasade commented:
Few comments, :+1: otherwise

antgonza commented:
@josenavas ping

RNAer commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,144 @@
+Transaction blocks are created through the same `TRN` variable exported in this
+module. You can add as many SQL commands as you want and execute all of them at
+once, and it will return the results of all the SQL commands. `TRN` should be
+used as a context so manager, and it autocommits the transaction once the last
```
"context so manager"?

josenavas replied:
Good catch! Fixed.

RNAer commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,144 @@
+used as a context so manager, and it autocommits the transaction once the last
+context is exited, as long as no error was generated inside the context, in
+which case a rollback is executed.
+
```
it is not clear to me what rollback is executed. It returns to the previous state before any added commands were run?

josenavas replied:
Yes, by definition of transaction, any command added to the current transaction is undone. Note that everything that has been already committed it will not be included in the rollback. Ex:

```python
with TRN:
    # (A) do stuff

with TRN:
    # (B) do stuff
    raise ValueError
```
In the above example, the stuff in (A) is committed, but the stuff in (B) is rollbacked. On the other hand:
```python
with TRN:
    with TRN:
        # (A) do stuff

    with TRN:
        # (B) do stuff
        raise ValueError
```
neither (A) or (B) are committed, so both of them will be rollbacked.

RNAer commented on qiita_db/sql_connection.py:
```diff
@@ -22,57 +22,144 @@
+>>> with TRN:
+...     sql = ("SELECT param_set_name, preprocessed_params_id "
+...            "FROM qiita.preprocessed_sequence_454_params "
+...            "WHERE barcode_type = %s ORDER BY preprocessed_params_id")
```
Doesn't this violate the SQL coding guidelines advocating the usage of `"""`?

josenavas replied:
You're correct. I removed this example because it is using the placeholders and we no longer need them.

josenavas commented:
Ok this is ready for another review. I removed the placeholders code, updated the portions of the code that were affected (only 2 functions) and added the Transaction to the CHANGELOG (it looks like the previous PR didn't got merged in master?)

antgonza commented:
:+1:

josenavas commented:
@RNAer or @wasade do you mind taking another look and merging if there are no more comments?

RNAer commented:
looking at it now

On Thu, Aug 13, 2015 at 9:20 PM, josenavas <notifications@github.com> wrote:

> @RNAer <https://github.com/RNAer> or @wasade <https://github.com/wasade>
> do you mind taking another look and merging if there are no more comments?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/qiita/pull/1407#issuecomment-130960444>.
>

RNAer commented:
:+1:

antgonza commented:
This has 2 :+1: so I'm gonna merge.
